### DISCUSSION TITLE
[Spring] Bean 객체와 Bean이 아닌 객체를 구별하는 방법

### DISCUSSION THREAD

**Discussion by SeonghaeJo** [wafflestudio/seminar-2023#152] [Q&A] [answered]
안녕하세요.
아래 코드에서 addArgumentResolvers로 파라미터 resolvers를 받는데, 제 생각에는 resolvers가 Bean으로 주입된 것 같습니다. (WebConfig 클래스 자체가 @Configuration 어노테이션이 붙어있으니까 Bean이고, resolvers가 Spring container 내에서 global & singleton으로 동작하는 것 같아 보이기 때문입니다.)
```Kotlin
// WebConfig.kt
@Configuration
class WebConfig(
    private val userArgumentResolver: UserArgumentResolver,
) : WebMvcConfigurer {

    override fun addArgumentResolvers(resolvers: MutableList<HandlerMethodArgumentResolver>) {
        resolvers.add(userArgumentResolver)
    }
}
```
반면에 아래 코드에서 authenticate의 파라미터인 accessToken은 Bean이 아닙니다.
```Kotlin
// UserServiceImpl.kt
@Service
class UserServiceImpl(
    private val userRepository: UserRepository,
) : UserService {
     override fun authenticate(accessToken: String): User {
        ...
    }
}
```
두 가지 경우 모두 Bean 객체의 메서드의 파라미터인데, resolvers는 Bean을 주입받는 것 같고 accessToken은 Bean이 아닌 것 같습니다.
이런 경우에 Bean인지 아닌지 어떻게 구분하나요? (@Autowired가 없다고 할 때)

**Comment by SeonghaeJo** [comment]
오른쪽에 Labels가 있긴 한데 제가 추가하거나 수정할 수 없는 것 같습니다. Reviewer 지정 못 하는 것처럼 권한 문제가 있는 건가요?

**Reply by PFCJeong** [reply]
네 이 이야기는 세미나장 채널에서도 이야기가 나왔는데요. 라벨은 세미나장 혹은 조교가 다는 걸로 결정하였습니다.

**Comment by PFCJeong** [comment, marked as answer]
안녕하세요~ 너무 좋은 주제네요

(1)

정확하지 않을 수 있지만, 결론만 말씀드리면 인자로 받은 resolvers는 빈이 아닙니다.

week1 세미나에서 `RequestHandlerMapping`를 다뤘었는데요. 정확히 설명은 안드렸는데 `DispatcherServlet`이 요청을 처리할 때 `RequestHandlerMapping` 말고도 `RequestHandlerMappingAdapter`를 사용합니다. (요 클래스는 pdf 18쪽 이미지에 등장하긴 합니다)

스프링에서 `RequestHandlerMappingAdapter`를 빈으로 만드는데요, `RequestHandlerMappingAdapter`는 프로퍼티로 argumentResolvers를 가지고 있습니다.

언급해주신 `resolvers`는 빈 리스트로 시작해서 여러 `addArgumentResolvers()`의 인자로 들어가서, 점차 채워집니다. 그리고 채워진 resolvers가 `RequestHandlerMappingAdapter`의 argumentResolvers 프로퍼티로 들어가게 됩니다.

(2)

`@Configuration`이 붙은 클래스의 함수에서 인자가 Bean임을 확인하는 방법은 간단한데요. 함수에 `@Bean` 어노테이션 여부가 곧 인자가 빈인지 아닌지를 결정합니다.

```
@Configuration
class ConfigA {
    
    @Bean
    fun beanB(a: BeanA): BeanB { // a는 스프링에서 관리하는 Bean이고, 이 인자는 스프링이 주입해주는 것
        
    }
    
    fun beanC(a: BeanA): BeanC { // a는 Bean일 수도 있고 Bean이 아닐 수도 있음. 사용하기 나름임. 단 이 인자는 스프링이 주입해주는 것이 아님
        
    }
}
```

도움이 되셨기를 바랍니다! (_ _)

**Comment by SeonghaeJo** [comment]
감사합니다!

정리하면 
1. 스프링이 시작할 때 `addArgumentResolvers`를 호출해서 resolvers를 채운 후, `RequestMappingHandlerAdapter`이라는 빈의 필드로 설정한다.
2. 빈 등록이 끝난 후 HTTP 요청이 오면 `RequestMappingHandlerAdapter` 빈을 사용한다.
(resolvers는 빈은 아니지만, 빈 객체의 필드이므로 global & singleton으로 동작할 수 있다)

인 것 같은데, 올바르게 이해한 것이 맞을까요?

추가로 답변의 예시 코드에서 `fun beanC(a: BeanA): BeanC`에서 beanC 메서드는 스프링이 시작할 때 자동으로 호출되는 것이 아니라 서비스 로직 상에서 직접 호출해야 하는 것이라 사용하기 나름인 것인가요?

**Reply by PFCJeong** [reply]
넵넵~ 정확하게 이해하셨습니다.
